File: checks/nvim-diff-zoom.lua
-- Configured Neovim: nvim --headless -c 'lua dofile("checks/nvim-diff-zoom.lua")'
local source = vim.env.NVIM_TEST_SOURCE
local function load(name)
	return source and dofile(source .. "/" .. name .. ".lua") or require("editor." .. name)
end
local Zoom, Diff = load("zellij-zoom"), load("git-diff")
local failures, passed = {}, 0
local function wait(predicate)
	assert(vim.wait(5000, predicate, 10), "timed out")
end
local function test(name, body)
	local ok, err = pcall(body)
	if ok then
		passed = passed + 1
		print("PASS: " .. name)
	else
		table.insert(failures, name .. ": " .. tostring(err))
	end
end
local function rig(options)
	local r = vim.tbl_extend(
		"force",
		{ fullscreen = false, toggles = 0, width = 60, calls = {}, warnings = {} },
		options or {}
	)
	if r.fullscreen then
		r.width = 120
	end
	r.zoom = Zoom.new({
		env = r.env or { ZELLIJ = "0", ZELLIJ_SESSION_NAME = "fixture session", ZELLIJ_PANE_ID = "7" },
		command = "/fixture/zellij",
		size = function()
			return r.width, 40
		end,
		notify = function(message)
			table.insert(r.warnings, message)
		end,
		run = function(argv, cb)
			assert(argv[1] == "/fixture/zellij" and argv[2] == "--session" and argv[3] == "fixture session")
			table.insert(r.calls, argv[5])
			vim.schedule(function()
				local failed = r.fail == argv[5]
				if failed then
					r.fail = nil
				end
				if argv[5] == "toggle-fullscreen" then
					assert(argv[6] == "--pane-id" and argv[7] == "terminal_7")
					if not failed or r.apply_then_fail then
						r.fullscreen = not r.fullscreen
						r.toggles = r.toggles + 1
						if not r.delay_resize then
							r.width = r.fullscreen and 120 or 60
						end
					end
				end
				local pane = {
					id = 7,
					is_plugin = false,
					is_fullscreen = r.fullscreen,
					is_floating = r.floating,
					pane_content_columns = r.fullscreen and 120 or 60,
					pane_content_rows = 40,
				}
				local panes = { { id = 7, is_plugin = true }, { id = 99, is_plugin = false } }
				if not r.missing then
					table.insert(panes, pane)
				end
				cb({
					code = failed and 1 or 0,
					stderr = failed and "fixture failure" or "",
					stdout = r.bad_json and "not json" or vim.json.encode(panes),
				})
			end)
		end,
	})
	function r.acquire()
		local complete, lease, err = false, nil, nil
		r.zoom.acquire(function(l, e)
			complete, lease, err = true, l, e
		end)
		wait(function()
			return complete
		end)
		return lease, err
	end
	return r
end

test("targeted fullscreen acquisition and restoration", function()
	local r = rig()
	local lease, err = r.acquire()
	assert(lease and not err and r.fullscreen and r.toggles == 1)
	r.zoom.release(lease)
	wait(function()
		return not r.fullscreen
	end)
	assert(r.toggles == 2)
end)
test("already-fullscreen panes stay fullscreen", function()
	local r = rig({ fullscreen = true })
	local lease = assert(r.acquire())
	r.zoom.release(lease)
	r.zoom.shutdown()
	assert(r.fullscreen and r.toggles == 0)
end)
test("outside Zellij runs no commands", function()
	local r = rig({ env = {} })
	r.zoom.release(assert(r.acquire()))
	r.zoom.shutdown()
	assert(#r.calls == 0)
end)
test("missing pane identity fails without commands", function()
	local r = rig({ env = { ZELLIJ = "0" } })
	local lease, err = r.acquire()
	assert(not lease and err and #r.calls == 0)
end)
for name, options in pairs({
	["query failure"] = { fail = "list-panes" },
	["invalid JSON"] = { bad_json = true },
	["missing pane"] = { missing = true },
	["floating pane"] = { floating = true },
	["toggle failure"] = { fail = "toggle-fullscreen" },
}) do
	test(name .. " does not grant a lease", function()
		local r = rig(options)
		local lease, err = r.acquire()
		assert(not lease and err and not r.fullscreen)
	end)
end
test("a toggle applied before failure is rolled back", function()
	local r = rig({ fail = "toggle-fullscreen", apply_then_fail = true })
	local lease, err = r.acquire()
	assert(not lease and err and not r.fullscreen and r.toggles == 2)
end)
test("waits for resize before granting lease", function()
	local r = rig({ delay_resize = true })
	local lease
	r.zoom.acquire(function(l, err)
		assert(not err)
		lease = l
	end)
	wait(function()
		return r.fullscreen
	end)
	assert(not vim.wait(80, function()
		return lease ~= nil
	end, 10))
	r.width = 120
	wait(function()
		return lease ~= nil
	end)
	r.zoom.shutdown()
end)
test("resize timeout rolls fullscreen back", function()
	local r = rig({ delay_resize = true })
	local lease, err = r.acquire()
	assert(not lease and err:find("Timed out") and not r.fullscreen)
end)
test("multiple users retain fullscreen until last release", function()
	local r = rig()
	local first, second = assert(r.acquire()), assert(r.acquire())
	r.zoom.release(first)
	assert(r.fullscreen and r.toggles == 1)
	r.zoom.release(second)
	wait(function()
		return not r.fullscreen
	end)
	assert(r.toggles == 2)
end)
test("manual unfullscreen is not toggled back on during release", function()
	local r = rig()
	local lease = assert(r.acquire())
	r.fullscreen = false
	r.zoom.release(lease)
	r.zoom.shutdown()
	assert(not r.fullscreen and r.toggles == 1)
end)
test("exit restores an active lease", function()
	local r = rig()
	assert(r.acquire())
	r.zoom.shutdown()
	assert(not r.fullscreen and r.toggles == 2)
end)

test("queued acquisitions serialize fullscreen toggles", function()
	local r = rig()
	local acquired = {}
	for _ = 1, 2 do
		r.zoom.acquire(function(lease, err)
			assert(not err)
			table.insert(acquired, lease)
		end)
	end
	wait(function()
		return #acquired == 2
	end)
	assert(r.toggles == 1)
	r.zoom.release(acquired[1])
	r.zoom.release(acquired[1])
	assert(r.fullscreen)
	r.zoom.release(acquired[2])
	r.zoom.shutdown()
	assert(not r.fullscreen and r.toggles == 2)
end)
test("exit cancels an in-flight acquisition", function()
	local r = rig()
	local finished, result, failure
	r.zoom.acquire(function(lease, err)
		finished, result, failure = true, lease, err
	end)
	r.zoom.shutdown()
	assert(finished and not result and failure and not r.fullscreen)
end)

-- Exercise the lifecycle with real Neovim windows and a controlled diff backend.
local original_notify = vim.notify
vim.notify = function() end
local serial = 0
local function diff_rig(options)
	options = options or {}
	pcall(vim.api.nvim_del_augroup_by_name, "EditorGitDiff")
	vim.cmd("silent! tabonly!")
	vim.cmd("silent! only!")
	vim.cmd.enew()
	vim.cmd.diffoff()
	local r = { opened = 0, acquired = 0, released = 0, source = vim.api.nvim_get_current_win() }
	r.manager = Diff.setup({
		zoom = {
			acquire = function(callback)
				r.acquired = r.acquired + 1
				r.resolve = function()
					callback(options.zoom_error and nil or {}, options.zoom_error)
				end
				if not options.delayed then
					vim.schedule(r.resolve)
				end
			end,
			release = function()
				r.released = r.released + 1
			end,
			shutdown = function() end,
		},
		diffthis = function(callback)
			r.opened = r.opened + 1
			if options.open_error then
				error("diff creation failed")
			end
			if options.no_diff then
				callback()
				return
			end
			local win = vim.api.nvim_get_current_win()
			vim.cmd.vnew()
			serial = serial + 1
			vim.api.nvim_buf_set_name(0, "gitsigns://fixture//HEAD/test-" .. serial)
			vim.cmd.diffthis()
			r.original = vim.api.nvim_get_current_win()
			vim.api.nvim_set_current_win(win)
			vim.cmd.diffthis()
			callback()
		end,
	})
	return r
end
test("duplicate presses await zoom, toggle closes only its own diff", function()
	local r = diff_rig({ delayed = true })
	vim.cmd.vnew()
	local unrelated = vim.api.nvim_get_current_win()
	vim.api.nvim_set_current_win(r.source)
	r.manager.toggle()
	r.manager.toggle()
	assert(r.acquired == 1 and r.opened == 0)
	r.resolve()
	assert(r.opened == 1 and vim.wo.diff)
	r.manager.toggle()
	assert(r.released == 1 and not vim.wo.diff and vim.api.nvim_win_is_valid(unrelated))
end)
test("fullscreen failure prevents diff opening", function()
	local r = diff_rig({ zoom_error = "fixture zoom failure" })
	r.manager.toggle()
	vim.wait(40, function()
		return false
	end)
	assert(r.opened == 0)
end)
for name, options in pairs({ ["failed diff"] = { open_error = true }, ["no diff produced"] = { no_diff = true } }) do
	test(name .. " releases zoom", function()
		local r = diff_rig(options)
		r.manager.toggle()
		wait(function()
			return r.released == 1
		end)
	end)
end
test("vanished source cancels a pending open", function()
	local r = diff_rig({ delayed = true })
	r.manager.toggle()
	vim.cmd.vnew()
	vim.api.nvim_win_close(r.source, true)
	r.resolve()
	assert(r.released == 1 and r.opened == 0)
end)
test("changed focus cancels a pending open", function()
	local r = diff_rig({ delayed = true })
	r.manager.toggle()
	vim.cmd.vnew()
	r.resolve()
	assert(r.released == 1 and r.opened == 0)
end)
test("manually closing original releases zoom", function()
	local r = diff_rig()
	r.manager.toggle()
	wait(function()
		return r.original ~= nil
	end)
	vim.api.nvim_win_close(r.original, true)
	wait(function()
		return r.released == 1
	end)
	assert(not vim.wo.diff)
end)
test("repurposed original window is preserved", function()
	local r = diff_rig()
	r.manager.toggle()
	wait(function()
		return r.original ~= nil
	end)
	vim.api.nvim_win_set_buf(r.original, vim.api.nvim_create_buf(true, false))
	wait(function()
		return r.released == 1
	end)
	assert(vim.api.nvim_win_is_valid(r.original))
end)
test("manual source close releases zoom", function()
	local r = diff_rig()
	r.manager.toggle()
	wait(function()
		return r.original ~= nil
	end)
	vim.api.nvim_win_close(r.source, true)
	wait(function()
		return r.released == 1
	end)
end)
test("two tabs keep independent diff lifetimes", function()
	local r = diff_rig()
	r.manager.toggle()
	wait(function()
		return r.opened == 1
	end)
	local first = vim.api.nvim_get_current_tabpage()
	vim.cmd.tabnew()
	r.manager.toggle()
	wait(function()
		return r.opened == 2
	end)
	r.manager.toggle()
	assert(r.released == 1)
	vim.api.nvim_set_current_tabpage(first)
	assert(vim.wo.diff)
	r.manager.toggle()
	assert(r.released == 2)
end)
test("closing a diff tab releases its lease", function()
	local r = diff_rig()
	vim.cmd.tabnew()
	r.manager.toggle()
	wait(function()
		return r.original ~= nil
	end)
	vim.cmd.tabclose()
	wait(function()
		return r.released == 1
	end)
end)
test("manual vimdiff is left alone", function()
	local r = diff_rig()
	vim.cmd.diffthis()
	r.manager.toggle()
	assert(r.acquired == 0 and vim.wo.diff)
end)
vim.notify = original_notify
if #failures > 0 then
	io.stderr:write(table.concat(failures, "\n") .. "\n")
	vim.cmd("cquit 1")
else
	print(string.format("PASS: %d diff/fullscreen scenarios", passed))
	vim.cmd("qa!")
end

File: checks/nvim-diff-zoom.py
"""Exercise the configured editor in a disposable, isolated Zellij session.

python checks/nvim-diff-zoom.py --nvim /path/to/configured/bin/nvim --zellij /path/to/zellij
No commands are sent to an existing user session.
"""

import argparse
import json
import os
import shutil
import subprocess
import tempfile
import time
from pathlib import Path


def wait_for(predicate, message, timeout=15):
    deadline = time.monotonic() + timeout
    while time.monotonic() < deadline:
        result = predicate()
        if result:
            return result
        time.sleep(0.05)
    raise AssertionError(message)


def run(args):
    with tempfile.TemporaryDirectory(prefix="nvim-diff-zoom-") as tmp:
        base = Path(tmp)
        runtime = base / "run"
        runtime.mkdir(mode=0o700)
        config = base / "config"
        config.mkdir()
        (config / "config.kdl").write_text(
            "show_startup_tips false\nshow_release_notes false\n"
            "session_serialization false\n"
        )
        env = dict(
            os.environ,
            XDG_RUNTIME_DIR=str(runtime),
            XDG_CACHE_HOME=str(base / "cache"),
            XDG_STATE_HOME=str(base / "state"),
            XDG_DATA_HOME=str(base / "data"),
            ZELLIJ_CONFIG_DIR=str(config),
            GIT_CONFIG_GLOBAL="/dev/null",
            GIT_CONFIG_NOSYSTEM="1",
        )
        for key in (
            "ZELLIJ",
            "ZELLIJ_SESSION_NAME",
            "ZELLIJ_PANE_ID",
            "ZELLIJ_CONFIG_FILE",
            "NVIM",
            "NVIM_ADDRESS",
            "DEV_NVIM_SOCKET",
        ):
            env.pop(key, None)
        repo = base / "repo with spaces"
        repo.mkdir()
        socket = base / "editor.sock"

        def command(argv, check=True):
            return subprocess.run(
                argv,
                cwd=repo,
                env=env,
                capture_output=True,
                text=True,
                timeout=20,
                check=check,
            )

        def action(*argv):
            return command(
                [args.zellij, "--session", "diff-fixture", "action", *argv]
            ).stdout

        def panes():
            return [
                p
                for p in json.loads(action("list-panes", "--json", "--all"))
                if not p["is_plugin"]
            ]

        def lua(expr):
            result = command(
                [
                    args.nvim,
                    "--server",
                    str(socket),
                    "--remote-expr",
                    "luaeval(" + json.dumps("vim.json.encode((" + expr + "))") + ")",
                ]
            )
            return json.loads(result.stdout)

        def execute(body):
            return lua("(function() " + body + "; return true end)()")

        def diff_count():
            return lua(
                "#vim.tbl_filter(function(w) return vim.wo[w].diff end, vim.api.nvim_list_wins())"
            )

        def toggle():
            execute('vim.fn.maparg(" gd", "n", false, true).callback()')

        command(["git", "init", "--quiet"])
        file = repo / "example.txt"
        file.write_text("first\noriginal\nlast\n")
        second_file = repo / "second.txt"
        second_file.write_text("another original\n")
        command(["git", "add", "example.txt", "second.txt"])
        command(
            [
                "git",
                "-c",
                "user.name=Fixture",
                "-c",
                "user.email=fixture@example.invalid",
                "-c",
                "commit.gpgsign=false",
                "commit",
                "--quiet",
                "-m",
                "fixture",
            ]
        )
        file.write_text("first\nstaged change\nlast\n")
        command(["git", "add", "example.txt"])
        file.write_text("first\nstaged change\nunstaged change\n")
        second_file.write_text("another changed file\n")
        layout = base / "layout.kdl"
        layout.write_text(
            'layout {\n tab name="fixture" {\n  pane split_direction="vertical" {\n'
            + "   pane command="
            + json.dumps(args.nvim)
            + " cwd="
            + json.dumps(str(repo))
            + ' { args "--listen" '
            + json.dumps(str(socket))
            + " "
            + json.dumps(str(file))
            + "; }\n"
            + "   pane command="
            + json.dumps(shutil.which("bash"))
            + ' { args "--noprofile" "--norc"; }\n'
            + "  }\n }\n}\n"
        )
        started = False
        try:
            command(
                [
                    args.zellij,
                    "--new-session-with-layout",
                    str(layout),
                    "attach",
                    "--create-background",
                    "diff-fixture",
                ]
            )
            started = True
            wait_for(socket.exists, "editor socket did not appear")
            wait_for(
                lambda: lua("vim.b.gitsigns_status_dict ~= nil"),
                "Gitsigns did not attach",
            )
            pane_id = int(lua('vim.env.ZELLIJ_PANE_ID:gsub("^terminal_", "")'))
            initial = panes()
            assert len(initial) == 2, initial
            sentinel = next(p for p in initial if p["id"] != pane_id)

            def editor():
                return next(p for p in panes() if p["id"] == pane_id)

            # Record state at the exact point the real Gitsigns diff is invoked.
            execute(
                'local gs = require("gitsigns"); local original = gs.diffthis; '
                "gs.diffthis = function(...) vim.g.fixture_call_count = (vim.g.fixture_call_count or 0) + 1; "
                'local result = vim.system({vim.g.editor_zellij_command, "--session", vim.env.ZELLIJ_SESSION_NAME, "action", "list-panes", "--json", "--all"}, {text=true}):wait(); '
                "assert(result.code == 0); "
                "for _, p in ipairs(vim.json.decode(result.stdout)) do "
                "if not p.is_plugin and tostring(p.id) == vim.env.ZELLIJ_PANE_ID then "
                "vim.g.fixture_order_ok = p.is_fullscreen and vim.o.columns == p.pane_content_columns and vim.o.lines == p.pane_content_rows "
                "end end; return original(...) end"
            )
            # An unrelated Neovim split must survive diff close.
            execute(
                'vim.cmd.split(); vim.g.fixture_unrelated = vim.api.nvim_get_current_win(); vim.cmd.wincmd("p")'
            )
            assert not editor()["is_fullscreen"]
            toggle()
            wait_for(lambda: diff_count() == 2, "diff did not open")
            assert lua("vim.g.fixture_order_ok"), "diff opened before fullscreen/resize"
            assert editor()["is_fullscreen"]
            originals = lua(
                'vim.tbl_map(function(w) return vim.api.nvim_buf_get_lines(vim.api.nvim_win_get_buf(w), 0, -1, false) end, vim.tbl_filter(function(w) return vim.api.nvim_buf_get_name(vim.api.nvim_win_get_buf(w)):match("^gitsigns://") ~= nil end, vim.api.nvim_list_wins()))'
            )
            assert originals == [["first", "original", "last"]], originals
            toggle()
            wait_for(lambda: not editor()["is_fullscreen"], "layout not restored")
            assert diff_count() == 0
            assert lua("vim.api.nvim_win_is_valid(vim.g.fixture_unrelated)")
            print(
                "PASS: targeted fullscreen and resize before HEAD diff; close restores layout and preserves unrelated split"
            )

            action("toggle-fullscreen", "--pane-id", f"terminal_{pane_id}")
            toggle()
            wait_for(lambda: diff_count() == 2, "pre-fullscreen diff did not open")
            toggle()
            wait_for(lambda: diff_count() == 0, "pre-fullscreen diff did not close")
            assert editor()["is_fullscreen"], "pre-existing fullscreen was lost"
            action("toggle-fullscreen", "--pane-id", f"terminal_{pane_id}")
            print("PASS: pre-existing fullscreen is preserved")

            toggle()
            wait_for(lambda: diff_count() == 2, "manual-close diff did not open")
            execute(
                'for _, w in ipairs(vim.api.nvim_list_wins()) do if vim.api.nvim_buf_get_name(vim.api.nvim_win_get_buf(w)):match("^gitsigns://") then vim.api.nvim_win_close(w, false) end end'
            )
            wait_for(
                lambda: not editor()["is_fullscreen"],
                "manual close did not restore layout",
            )
            print("PASS: manual diff-window close restores layout")

            # Two Neovim tabs share one containing Zellij pane lease.
            toggle()
            wait_for(lambda: diff_count() == 2, "first tab diff did not open")
            execute("vim.cmd.tabnew(" + json.dumps(str(second_file)) + ")")
            wait_for(
                lambda: lua("vim.b.gitsigns_status_dict ~= nil"),
                "second tab did not attach",
            )
            toggle()
            wait_for(lambda: diff_count() == 4, "second tab diff did not open")
            toggle()
            wait_for(lambda: diff_count() == 2, "second tab diff did not close")
            assert editor()["is_fullscreen"], (
                "closing one tab released another tab's zoom"
            )
            execute("vim.cmd.tabclose()")
            toggle()
            wait_for(
                lambda: not editor()["is_fullscreen"], "last diff did not release zoom"
            )
            print("PASS: multiple diff tabs share fullscreen until the last closes")

            # With no Gitsigns attachment, opening is a no-op and must roll back.
            execute("vim.cmd.enew()")
            count = lua("vim.g.fixture_call_count")
            toggle()
            wait_for(
                lambda: lua("vim.g.fixture_call_count") > count,
                "failure-path diff did not run",
            )
            wait_for(
                lambda: not editor()["is_fullscreen"], "no-op diff did not roll back"
            )
            print("PASS: a buffer with no comparison rolls fullscreen back")

            after = next(p for p in panes() if p["id"] == sentinel["id"])
            assert (after["tab_id"], after["terminal_command"]) == (
                sentinel["tab_id"],
                sentinel["terminal_command"],
            )
            execute("vim.cmd.edit(" + json.dumps(str(file)) + ")")
            toggle()
            wait_for(lambda: diff_count() == 2, "exit-test diff did not open")
            execute('vim.schedule(function() vim.cmd("qa!") end)')
            wait_for(lambda: not socket.exists(), "editor did not exit")
            assert not any(p["is_fullscreen"] for p in panes())
            print("PASS: editor exit cleans up; unrelated pane untouched")
        except Exception:
            if socket.exists():
                print(
                    "Fixture diagnostics:",
                    lua(
                        '(function() local result = {}; for _, w in ipairs(vim.api.nvim_list_wins()) do table.insert(result, {window=w, tab=vim.api.nvim_win_get_tabpage(w), diff=vim.wo[w].diff, buffer=vim.api.nvim_buf_get_name(vim.api.nvim_win_get_buf(w))}) end; return {windows=result, messages=vim.api.nvim_exec2("messages", {output=true}).output} end)()'
                    ),
                )
            raise
        finally:
            if started:
                command([args.zellij, "kill-session", "diff-fixture"], check=False)


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("--nvim", required=True)
    parser.add_argument("--zellij", default=shutil.which("zellij"))
    run(parser.parse_args())
